disk-buffering/src/main/java/io/opentelemetry/contrib/disk/buffering/internal/exporters/AbstractDiskExporter.java

```diff
@@ -0,0 +1,91 @@
+import io.opentelemetry.contrib.disk.buffering.internal.storage.Storage;
+import io.opentelemetry.contrib.disk.buffering.storage.StorageConfiguration;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public abstract class AbstractDiskExporter<EXPORT_DATA> {
+  private final Storage storage;
+  private static final Logger logger = Logger.getLogger(AbstractDiskExporter.class.getName());
+
+  public AbstractDiskExporter(File rootDir, StorageConfiguration configuration) {
+    validateConfiguration(configuration);
+    this.storage = new Storage(new FolderManager(getSignalFolder(rootDir), configuration));
+  }
+
+  /**
+   * Reads data from the disk and attempts to export it.
+   *
+   * @param timeout The amount of time to wait for the wrapped exporter to finish.
+   * @param unit The unit of the time provided.
+   * @return TRUE if there was data available and it was successfully exported within the timeout
+   *     provided. FALSE if either of those conditions didn't meet.
+   * @throws IOException If an unexpected error happens.
+   */
+  public boolean exportStoredBatch(long timeout, TimeUnit unit) throws IOException {
+    logger.log(Level.INFO, "Attempting to export batch from disk.");
+    AtomicBoolean exportSucceeded = new AtomicBoolean(false);
+    boolean foundDataToExport =
+        storage.read(
+            bytes -> {
+              logger.log(Level.INFO, "About to export stored batch.");
+              CompletableResultCode join =
+                  doExport(getSerializer().deserialize(bytes)).join(timeout, unit);
+              exportSucceeded.set(join.isSuccess());
+              return exportSucceeded.get();
+            });
+    return foundDataToExport && exportSucceeded.get();
+  }
+
+  protected abstract String getStorageFolderName();
+
+  protected abstract CompletableResultCode doExport(Collection<EXPORT_DATA> data);
+
+  protected abstract SignalSerializer<EXPORT_DATA> getSerializer();
+
+  protected void onShutDown() throws IOException {
+    storage.close();
+  }
+
+  protected CompletableResultCode onExport(Collection<EXPORT_DATA> data) {
+    logger.log(Level.DEBUG, "Intercepting exporter batch.");
+    try {
+      storage.write(getSerializer().serialize(data));
+      return CompletableResultCode.ofSuccess();
+    } catch (IOException e) {
+      logger.log(Level.INFO, "Could not store batch in disk. Exporting it right away.");
+      return doExport(data);
+    }
+  }
+
+  private File getSignalFolder(File rootDir) {
+    File folder = new File(rootDir, getStorageFolderName());
+    if (!folder.exists()) {
+      if (!folder.mkdirs()) {
+        throw new IllegalStateException("Could not create the signal folder");
```

Comment:
is throwing an exception (and thus crashing the app) appropriate? can we just return an optional instead?

Reply:
I missed this comment for the previous changes I made. So from a consumer point of view, this code will be executed when creating a new instance of either of the signal disk exporters. So maybe, in order to return some value instead of throwing an exception, we might have to make the constructors private and use some static function for creating the instances instead. Apart from that, I'm just realizing that this exception isn't present in the constructor's signature, so there's no compilation warning that people can see to become aware of this possible scenario, so that's an issue.

Now, I think there has to be some sort of specific feedback for this because, if the folder cannot be created, it's most likely due to some issue with the `rootDir` File provided by the consumer, so my concern with using an optional, or even just returning null when creating an instance through a static function, would be that the reason why the class couldn't be instantiated wouldn't be clear, so the consumer wouldn't easily know what to fix, which they have to I believe since they should be aware of the disk space and permissions available for the folder they provide. Though I'm open to suggestions to achieve the same with another mechanism. Now, if the exception is what we end up choosing though, I think I'll have to convert it into an IOException one to provide the compile-time checks.

Reply:
sounds good :smile:

Reply:
I've decided to create a static `create` function for each disk exporter since it seems to be a common pattern in other classes of the Java SDK.